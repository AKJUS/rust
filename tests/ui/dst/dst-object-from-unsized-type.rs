// Test that we cannot create objects from unsized types.
//@ revisions: current next
//@ ignore-compare-mode-next-solver (explicit revisions)
//@[next] compile-flags: -Znext-solver

trait Foo { fn foo(&self) {} }
impl Foo for str {}
impl Foo for [u8] {}

fn test1<T: ?Sized + Foo>(t: &T) {
    let u: &dyn Foo = t;
    //~^ ERROR the size for values of type
}

fn test2<T: ?Sized + Foo>(t: &T) {
    let v: &dyn Foo = t as &dyn Foo;
    //~^ ERROR the size for values of type
}

fn test3() {
    let _: &[&dyn Foo] = &["hi"];
    //~^ ERROR the size for values of type
}

fn test4(x: &[u8]) {
    let _: &dyn Foo = x as &dyn Foo;
    //~^ ERROR the size for values of type
}

fn main() { }
