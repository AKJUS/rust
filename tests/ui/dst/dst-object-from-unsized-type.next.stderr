error[E0308]: mismatched types
  --> $DIR/dst-object-from-unsized-type.rs:11:23
   |
LL | fn test1<T: ?Sized + Foo>(t: &T) {
   |          - found this type parameter
LL |     let u: &dyn Foo = t;
   |            --------   ^ expected `&dyn Foo`, found `&T`
   |            |
   |            expected due to this
   |
   = note: expected reference `&dyn Foo`
              found reference `&T`
   = help: type parameters must be constrained to match other types
   = note: for more information, visit https://doc.rust-lang.org/book/ch10-02-traits.html#traits-as-parameters

error[E0605]: non-primitive cast: `&T` as `&dyn Foo`
  --> $DIR/dst-object-from-unsized-type.rs:16:23
   |
LL |     let v: &dyn Foo = t as &dyn Foo;
   |                       ^^^^^^^^^^^^^ an `as` expression can only be used to convert between primitive types or to coerce to a specific trait object

error[E0308]: mismatched types
  --> $DIR/dst-object-from-unsized-type.rs:21:28
   |
LL |     let _: &[&dyn Foo] = &["hi"];
   |                            ^^^^ expected `&dyn Foo`, found `&str`
   |
   = note: expected reference `&dyn Foo`
              found reference `&'static str`
   = help: `str` implements `Foo` so you could box the found value and coerce it to the trait object `Box<dyn Foo>`, you will have to change the expected type as well

error[E0605]: non-primitive cast: `&[u8]` as `&dyn Foo`
  --> $DIR/dst-object-from-unsized-type.rs:26:23
   |
LL |     let _: &dyn Foo = x as &dyn Foo;
   |                       ^^^^^^^^^^^^^ an `as` expression can only be used to convert between primitive types or to coerce to a specific trait object

error: aborting due to 4 previous errors

Some errors have detailed explanations: E0308, E0605.
For more information about an error, try `rustc --explain E0308`.
